dataset:

|  x1  |  x2  | Class |
| :--: | :--: | :---: |
|  -2  |    1 |    A  |
|  -1  |    2 |    A  |
|   1  |    3 |    A  |
|   2  |    4 |    A  |
|  -1  |  0.4 |    B  |
|   2  |  1.1 |    B  |
|   3  |  1.5 |    B  |
|   4  |  1.8 |    B  |

query vector = [1,2] ^ T (transposed)

In [25]:
import sys
import os
import importlib

#Add code paths
sys.path.append(os.path.abspath("../code/part1/"))


In [26]:
import numpy as np

x1 = np.array([-2,-1,1,2,-1,2,3,4])
x2 = np.array([1,2,3,4,0.4,1.1,1.5,1.8])

classes = np.array(["A","A","A","A","B","B","B","B"])

query = np.array([1,2])

a) Compute the most probable class for the query vector, under the Naive Bayes assumption, using 1-dimensional Gaussians to model the likelihoods.

In [27]:
import bayesian_a as b1
importlib.reload(b1) 

b1.calculate_posteriors_1d(x1,x2,classes,query,debug=False)


Class A
SCORE: 0.03334853656931131
Class B
SCORE: 0.02196275648871398


b) Compute the most probable class for the query vector assuming that the likelihoods are 2-dimensional Gaussians. Are the results from 1 a) and 1 b) the same? Why are the same or not (one sentence, no mathematical proof required) 

In [28]:
import bayesian_b as b2
importlib.reload(b2) 

b2.calculate_posteriors_2d(x1,x2,classes,query,debug=False) 

# ignore the list that is being returned -> its because of d)

Class A
RESULT: 1.0
Class B
RESULT: 8.311619269398749e-47


[np.float64(1.0), np.float64(8.311619269398749e-47)]

JUSTIFICATION

| EXERCISE |             CLASS A |               CLASS B | RESULT |
| :------: | :-----------------: | :-------------------: | :----: |
|        1 | 0.03334853656931131 |   0.02196275648871398 |      A |
|        2 |                   1 | 8.311619269398749e-47 |      A |

Both approaches identify Class A as the most probable, but the posterior probabilities differ because the multivariate Gaussian in 1 a) accounts for correlations between features, whereas the Naive Bayes 1D Gaussian model in 1 b) assumes feature independence.

The results from 1 a) and 1 b) are the same, both predict class A. This is because, although the likelihoods are modelled differently (1D vs 2D Gaussians), the combined probability still favours class A in both cases, so the most probable class does not change.

dataset:

|  x3  | Class |
| :--: | :---: |
|   0  |    A  |
|   1  |    A  |
|   1  |    A  |
|   0  |    A  |
|   1  |    B  |
|   1  |    B  |
|   0  |    B  |
|   1  |    B  |

query vector = True = 1 

1 = True; 0 = False

In [29]:
x3 = np.array([0,1,1,0,1,1,0,1])

classes = np.array(["A","A","A","A","B","B","B","B"])

query = np.array([1])

c) Compute the most probable class, with x3 being a categorial class 1=True, 0=False.

In [30]:
import bayesian_c as b3
importlib.reload(b3) 

b3.calculate_most_probable(classes, x3, query, debug = False)

Class A
RESULT: 0.4
Class B
RESULT: 0.6
MOST PROBABLE CLASS: B with prob of 0.6


dataset:

|  x1  |  x2  |  x3  | Class |
| :--: | :--: | :--: | :---: |
|  -2  |    1 |    0 |    A  |
|  -1  |    2 |    1 |    A  |
|   1  |    3 |    1 |    A  |
|   2  |    4 |    0 |    A  |
|  -1  |  0.4 |    1 |    B  |
|   2  |  1.1 |    1 |    B  |
|   3  |  1.5 |    0 |    B  |
|   4  |  1.8 |    1 |    B  |

x1 and x2 are dependable
x3 is independent of x1 and x2. 

x3 is a categorial class: 1 = True; 0 = False

query vector = (1,2,1) 


In [31]:
x1 = np.array([-2,-1,1,2,-1,2,3,4])
x2 = np.array([1,2,3,4,0.4,1.1,1.5,1.8])
x3 = np.array([0,1,1,0,1,1,0,1])

classes = np.array(["A","A","A","A","B","B","B","B"])

query = np.array([1,2,1])

d) Compute the most probable class and indicate the estimated relative probability.

In [32]:
import bayesian_d as b4
importlib.reload(b4) 

b4.calculate_most_probable(classes, x1, x2, x3, query, debug = False)

Class A
RESULT: 1.0
Class B
RESULT: 1.2467428904098125e-46
